src/lib/orionld/serviceRoutines/orionldPostBatchUpdate.cpp

```diff
@@ -0,0 +1,250 @@
+#include "orionld/kjTree/kjStringValueLookupInArray.h"         // kjStringValueLookupInArray
+#include "orionld/kjTree/kjTreeToUpdateContextRequest.h"       // kjTreeToUpdateContextRequest
+#include "orionld/serviceRoutines/orionldPostBatchUpdate.h"    // Own Interface
+
+
+// ----------------------------------------------------------------------------
+//
+// entityIdPush - add ID to array
+//
+static void entityIdPush(KjNode* entityIdsArrayP, const char* entityId)
+{
+  KjNode* idNodeP = kjString(orionldState.kjsonP, NULL, entityId);
+
+  kjChildAdd(entityIdsArrayP, idNodeP);
+}
+
+
+// -----------------------------------------------------------------------------
+//
+// entityIdGet -
+//
+// static void entityIdGet(KjNode* dbEntityP, char** idP)
+// {
+//   for (KjNode* nodeP = dbEntityP->value.firstChildP; nodeP != NULL; nodeP = nodeP->next)
+//   {
+//     if (SCOMPARE3(nodeP->name, 'i', 'd', 0))
+//       *idP = nodeP->value.s;
+//   }
+// }
+
+
+// ----------------------------------------------------------------------------
+//
+// entitySuccessPush -
+//
+static void entitySuccessPush(KjNode* successArrayP, const char* entityId)
+{
+  KjNode* eIdP = kjString(orionldState.kjsonP, "id", entityId);
+
+  kjChildAdd(successArrayP, eIdP);
+}
+
+
+// ----------------------------------------------------------------------------
+//
+// orionldPostEntityOperationsUpdate -
+//
+// POST /ngsi-ld/v1/entityOperations/update
+//
+// From the spec:
+//   This operation allows updating a batch of NGSI-LD Entities.
+//
+bool orionldPostBatchUpdate(ConnectionInfo* ciP)
+{
+  //
+  // Prerequisites for the payload in orionldState.requestTree:
+  // * must be an array
+  // * cannot be empty
+  // * all entities must contain a entity::id (one level down)
+  // * no entity can contain an entity::type (one level down)
+  //
+  pcheckEntities(orionldState.requestTree);
+
+  KjNode*               incomingTree   = orionldState.requestTree;
+  KjNode*               idArray        = kjArray(orionldState.kjsonP, NULL);
+  KjNode*               successArrayP  = kjArray(orionldState.kjsonP, "success");
+  KjNode*               errorsArrayP   = kjArray(orionldState.kjsonP, "errors");
+  KjNode*               entityP;
+  KjNode*               next;
+
+  //
+  // 01. Create idArray as an array of entity IDs, extracted from orionldState.requestTree
+  //
+  entityP = incomingTree->value.firstChildP;
+  while (entityP)
+  {
+    next = entityP->next;
+
+    char* entityId;
+    char* entityType;
+
+    // entityIdAndTypeGet calls entityIdCheck/entityTypeCheck that adds the entity in errorsArrayP if needed
+    if (entityIdAndTypeGet(entityP, &entityId, &entityType, errorsArrayP) == true)
+      entityIdPush(idArray, entityId);
+    else
+      kjChildRemove(incomingTree, entityP);
+
+    entityP = next;
+  }
+
+  //
+  // 02. Check if some ID from idArray not exists
+  //
+  for (KjNode* idEntity = idArray->value.firstChildP; idEntity != NULL; idEntity = idEntity->next)
```

Comment:
Don't we have a function to do this check with a single access to mongo?
If don't have one, we need to implement one.
Can't perform one DB access per entity ... way too slow

Reply:
All right, Ken